pkg/util/pods.go

```diff
@@ -134,79 +138,68 @@ func WaitForPodCompletion(ctx context.Context, podClient kubernetes.PodClient, n
 }
 
 func waitForPodCompletionOrTimeout(ctx context.Context, podClient kubernetes.PodClient, namespace, name string, completed map[string]time.Time, notifier ContainerNotifier, skipLogs bool) (*corev1.Pod, error) {
-	// Warning: this is extremely fragile, inherited legacy code.  Please be
-	// careful and test thoroughly when making changes, as they very frequently
-	// lead to systemic production failures.  Some guidance:
-	// - There is a complex interaction between this code and the container
-	//   notifier.  Updates to the state of the pod are received via the watch
-	//   and communicated to the notifier.  Even in case of interruption (i.e.
-	//   cancellation of `ctx`) and/or failure, events should continue to be
-	//   processed until the notifier signals that it is done.  This ensures
-	//   the state of the pod is correctly reported, artifacts are gathered,
-	//   and termination happens deterministically for both success and failure
-	//   scenarios.
-	// - Since ea8f62fcf, most of the above only applies to template tests.
-	//   Container and multi-stage tests now solely rely on `test-infra`'s
-	//   `pod-utils` for artifact gathering and so use a notifier which
-	//   instantly reports itself as done when the watched containers finish.
-	pod := &corev1.Pod{}
-	if err := podClient.Get(ctx, ctrlruntimeclient.ObjectKey{Namespace: namespace, Name: name}, pod); err != nil {
-		if kerrors.IsNotFound(err) {
-			notifier.Complete(name)
-			logrus.Infof("error: could not wait for pod '%s': it is no longer present on the cluster"+
-				" (usually a result of a race or resource pressure. re-running the job should help)", name)
-			return nil, fmt.Errorf("pod was deleted while ci-operator step was waiting for it")
+	var ret atomic.Pointer[corev1.Pod]
+	var eg *errgroup.Group
+	eg, ctx = errgroup.WithContext(ctx)
+	pendingCtx, cancel := context.WithCancel(ctx)
+	pendingCheck := func() error {
+		timeout := podClient.PendingTimeout()
+		if pod, err := checkPendingPeriodic(pendingCtx.Done(), timeout, &ret); err != nil {
+			err = fmt.Errorf("pod pending for more than %s: %w: %s\n%s", timeout, err, getReasonsForUnreadyContainers(pod), getEventsForPod(ctx, pod, podClient))
+			logrus.Info(err)
+			notifier.Complete(pod.Name)
+			return err
 		}
-		return nil, fmt.Errorf("could not list pod: %w", err)
+		return nil
 	}
+	eg.Go(func() error {
+		defer cancel()
+		if err := kubernetes.WaitForConditionOnObject(ctx, podClient, ctrlruntimeclient.ObjectKey{Namespace: namespace, Name: name}, &corev1.PodList{}, &corev1.Pod{}, func(obj runtime.Object) (bool, error) {
+			pod := obj.(*corev1.Pod)
+			if ret.Swap(pod) == nil {
```

Comment:
Yeah, there's a lot in these few lines.  I've added a description of the implementation in general as a comment and in the commit message (https://github.com/openshift/ci-tools/pull/3281/commits/bae1a6750f2298e4a6900d957d73585339abd875).

The intent of this particular condition is to start the Go routine once when the first state is seen (since we need a least one pod object to perform it).  The `nil` value (initialized by the `var` statement at the top of the function) is used for that purpose.  `Swap` is not strictly necessary but, as mentioned in the commit message, Go does not provide weaker primitives and it makes the code a bit shorter.